File: android/src/main/kotlin/com/techlabs/flutter/flutter_techlabs_library/FlutterTechlabsLibraryPlugin.kt
package com.techlabs.flutter.flutter_techlabs_library

import com.techlabs.flutter.flutter_techlabs_library.lib.ServerEnvironment
import com.techlabs.flutter.flutter_techlabs_library.lib.TechLabsLibrary
import com.techlabs.flutter.flutter_techlabs_library.lib.model.NoticeCheckResult
import com.techlabs.flutter.flutter_techlabs_library.lib.model.VersionCheckResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class FlutterTechlabsLibraryPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private var scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // 재연결 시 새 scope 생성 (이전 scope가 cancel된 경우 대비)
        scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
        channel = MethodChannel(binding.binaryMessenger, "com.techlabs.flutter/library")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }

    @Suppress("UNCHECKED_CAST")
    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "configure" -> {
                val appId = call.argument<String>("appId") ?: run {
                    result.error("INVALID_ARGUMENT", "appId is required", null)
                    return
                }
                val envStr = call.argument<String>("environment") ?: "production"
                val environment = if (envStr == "development") ServerEnvironment.DEVELOPMENT
                else ServerEnvironment.PRODUCTION
                // Flutter int는 64비트이므로 Long으로 받아 Int로 변환
                val connectTimeoutMs = (call.argument<Any>("connectTimeoutMs") as? Number)?.toInt() ?: 10000
                val readTimeoutMs = (call.argument<Any>("readTimeoutMs") as? Number)?.toInt() ?: 10000
                TechLabsLibrary.init(appId, environment, connectTimeoutMs, readTimeoutMs)
                result.success(null)
            }

            "fetchServiceInfo" -> {
                scope.launch {
                    try {
                        val info = TechLabsLibrary.fetchServiceInfoSuspend()
                        if (isActive) result.success(serviceInfoToMap(info))
                    } catch (e: kotlinx.coroutines.CancellationException) {
                        throw e
                    } catch (e: Exception) {
                        if (isActive) result.error("FETCH_ERROR", e.message, null)
                    }
                }
            }

            "checkVersion" -> {
                val infoMap = call.argument<Map<String, Any?>>("info")
                val currentVersion = call.argument<String>("currentVersion")
                val currentBuildNumber = (call.argument<Any>("currentBuildNumber") as? Number)?.toInt()

                val versionResult = when {
                    currentBuildNumber != null -> TechLabsLibrary.checkVersion(infoMap, currentBuildNumber)
                    currentVersion != null -> TechLabsLibrary.checkVersion(infoMap, currentVersion)
                    else -> VersionCheckResult.Error
                }
                result.success(versionCheckResultToMap(versionResult))
            }

            "checkNotice" -> {
                val infoMap = call.argument<Map<String, Any?>>("info")
                val lastSeenIndex = (call.argument<Any>("lastSeenIndex") as? Number)?.toInt() ?: 0
                val noticeResult = TechLabsLibrary.checkNotice(infoMap, lastSeenIndex)
                result.success(noticeCheckResultToMap(noticeResult))
            }

            "clearCache" -> {
                TechLabsLibrary.clearCache()
                result.success(null)
            }

            "isUpdateRequired" -> {
                val serverVersion = call.argument<String>("serverVersion")
                val currentVersion = call.argument<String>("currentVersion")
                val serverBuildNumber = (call.argument<Any>("serverBuildNumber") as? Number)?.toInt()
                val currentBuildNumber = (call.argument<Any>("currentBuildNumber") as? Number)?.toInt()

                val isRequired = when {
                    serverBuildNumber != null && currentBuildNumber != null ->
                        TechLabsLibrary.isUpdateRequired(serverBuildNumber, currentBuildNumber)
                    serverVersion != null && currentVersion != null ->
                        TechLabsLibrary.isUpdateRequired(serverVersion, currentVersion)
                    else -> false
                }
                result.success(isRequired)
            }

            else -> result.notImplemented()
        }
    }

    private fun serviceInfoToMap(info: com.techlabs.flutter.flutter_techlabs_library.lib.model.ServiceInfo): Map<String, Any?> {
        return mapOf(
            "domain" to info.domain,
            "maintenance" to info.maintenance,
            "message" to info.message,
            "title" to info.title,
            "appAction" to info.appAction,
            "appVerInfo" to info.appVerInfo.map { a ->
                mapOf(
                    "device" to a.device,
                    "app_name" to a.appName,
                    "ver" to a.ver,
                    "type" to a.type,
                    "app_name2" to a.appName2,
                    "latest_ver" to a.latestVer
                )
            }
        )
    }

    private fun versionCheckResultToMap(result: VersionCheckResult): Map<String, Any?> {
        return when (result) {
            is VersionCheckResult.ForceUpdate -> mapOf(
                "result" to "forceUpdate",
                "storePackage" to result.storePackage,
                "callbackCode" to 1
            )
            is VersionCheckResult.OptionalUpdate -> mapOf(
                "result" to "optionalUpdate",
                "storePackage" to result.storePackage,
                "callbackCode" to 2
            )
            VersionCheckResult.NoUpdateNeeded -> mapOf(
                "result" to "noUpdateNeeded",
                "storePackage" to null,
                "callbackCode" to 0
            )
            VersionCheckResult.Error -> mapOf(
                "result" to "error",
                "storePackage" to null,
                "callbackCode" to 3
            )
        }
    }

    private fun noticeCheckResultToMap(result: NoticeCheckResult): Map<String, Any?> {
        return when (result) {
            is NoticeCheckResult.HasNewNotice -> mapOf(
                "result" to "hasNewNotice",
                "latestIndex" to result.latestIndex
            )
            NoticeCheckResult.NoNewNotice -> mapOf(
                "result" to "noNewNotice",
                "latestIndex" to null
            )
            NoticeCheckResult.Error -> mapOf(
                "result" to "error",
                "latestIndex" to null
            )
        }
    }
}

File: ios/flutter_techlabs_library/Sources/flutter_techlabs_library/FlutterTechlabsLibraryPlugin.swift
import Flutter
import UIKit
import ios_techlabs_library

public class FlutterTechlabsLibraryPlugin: NSObject, FlutterPlugin {

    private var methodChannel: FlutterMethodChannel?
    private var inflightTasks: [AnyObject] = []

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.techlabs.flutter/library",
            binaryMessenger: registrar.messenger()
        )
        let instance = FlutterTechlabsLibraryPlugin()
        instance.methodChannel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        for t in inflightTasks {
            (t as? Task<Void, Never>)?.cancel()
        }
        inflightTasks.removeAll()
        methodChannel?.setMethodCallHandler(nil)
        methodChannel = nil
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":
            handleConfigure(call: call, result: result)
        case "fetchServiceInfo":
            handleFetchServiceInfo(result: result)
        case "checkVersion":
            handleCheckVersion(call: call, result: result)
        case "checkNotice":
            handleCheckNotice(call: call, result: result)
        case "clearCache":
            TechLabsLibrary.shared.clearCache()
            result(nil)
        case "isUpdateRequired":
            handleIsUpdateRequired(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleConfigure(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let appId = args["appId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "appId is required", details: nil))
            return
        }
        let envStr = args["environment"] as? String ?? "production"
        let environment: ServerEnvironment = envStr == "development" ? .development : .production
        TechLabsLibrary.shared.configure(appIdentifier: appId, environment: environment)
        if TechLabsLibrary.shared.appIdentifier == nil {
            result(FlutterError(code: "INVALID_APP_ID", message: "Invalid appId: \(appId)", details: nil))
            return
        }
        result(nil)
    }

    private func handleFetchServiceInfo(result: @escaping FlutterResult) {
        if #available(iOS 15.0, *) {
            let task: Task<Void, Never> = Task {
                do {
                    let info = try await TechLabsLibrary.shared.fetchServiceInfo()
                    guard !Task.isCancelled else { return }
                    await MainActor.run {
                        result(self.serviceInfoToMap(info))
                    }
                } catch {
                    guard !Task.isCancelled else { return }
                    await MainActor.run {
                        result(FlutterError(code: "FETCH_ERROR", message: "Service info fetch failed", details: nil))
                    }
                }
            }
            inflightTasks.append(task)
        } else {
            result(FlutterError(code: "UNSUPPORTED", message: "iOS 15.0+ required", details: nil))
        }
    }

    private func handleCheckVersion(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "arguments required", details: nil))
            return
        }
        guard let infoDict = args["info"] as? [String: Any] else {
            result(versionCheckResultToMap(.error))
            return
        }
        let info = VersionCheckInfo(dictionary: infoDict)
        let currentBuildNumber = (args["currentBuildNumber"] as? NSNumber)?.intValue
        let currentVersion = args["currentVersion"] as? String

        if let buildNum = currentBuildNumber {
            guard !info.ver.isEmpty else {
                result(versionCheckResultToMap(.noUpdateNeeded))
                return
            }
            guard let serverBuild = Int(info.ver) else {
                result(versionCheckResultToMap(.noUpdateNeeded))
                return
            }
            let needs = serverBuild > buildNum
            result(versionCheckResultToMap(resolveResultLocal(isUpdateRequired: needs, info: info)))
            return
        }

        guard let currentVersion = currentVersion else {
            result(versionCheckResultToMap(.error))
            return
        }
        let versionResult = TechLabsLibrary.shared.checkVersion(info: info, currentVersion: currentVersion)
        result(versionCheckResultToMap(versionResult))
    }

    private func resolveResultLocal(isUpdateRequired: Bool, info: VersionCheckInfo) -> VersionCheckResult {
        guard isUpdateRequired else { return .noUpdateNeeded }
        switch info.type {
        case "0": return .forceUpdate(storePackage: info.appName2)
        case "1": return .optionalUpdate(storePackage: info.appName2)
        default: return .error
        }
    }

    private func handleCheckNotice(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "arguments required", details: nil))
            return
        }
        guard let infoDict = args["info"] as? [String: Any] else {
            result(noticeCheckResultToMap(.error))
            return
        }
        let lastSeenIndex = args["lastSeenIndex"] as? Int ?? 0

        let noticeResult = TechLabsLibrary.shared.checkNotice(
            dictionary: infoDict,
            lastSeenIndex: lastSeenIndex
        )
        result(noticeCheckResultToMap(noticeResult))
    }

    private func handleIsUpdateRequired(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(false)
            return
        }
        let serverVersion = args["serverVersion"] as? String
        let currentVersion = args["currentVersion"] as? String
        // iOS는 빌드넘버(정수) 비교를 지원하지 않으므로 버전 문자열로 비교
        // serverBuildNumber/currentBuildNumber가 넘어오더라도 문자열로 변환해 처리
        let serverBuildNumber = (args["serverBuildNumber"] as? NSNumber).map { "\($0.intValue)" }
        let currentBuildNumber = (args["currentBuildNumber"] as? NSNumber).map { "\($0.intValue)" }

        if let sv = serverVersion, let cv = currentVersion {
            result(TechLabsLibrary.shared.isUpdateRequired(serverVersion: sv, currentVersion: cv))
        } else if let sv = serverBuildNumber, let cv = currentBuildNumber {
            // 빌드넘버를 정수 문자열로 비교 (큰 숫자가 더 최신)
            let serverInt = Int(sv) ?? 0
            let currentInt = Int(cv) ?? 0
            result(serverInt > currentInt)
        } else {
            result(false)
        }
    }

    // MARK: - Map helpers

    private func serviceInfoToMap(_ info: ServiceInfo) -> [String: Any] {
        [
            "domain": info.domain,
            "maintenance": info.maintenance,
            "message": info.message,
            "title": info.title,
            "appAction": info.appAction,
            "appVerInfo": info.appVerInfo.map(appVerInfoToMap)
        ]
    }

    private func appVerInfoToMap(_ a: AppVerInfo) -> [String: Any] {
        [
            "device": a.device,
            "app_name": a.appName,
            "ver": a.ver,
            "type": a.type,
            "app_name2": a.appName2,
            "latest_ver": a.latestVer
        ]
    }

    private func versionCheckResultToMap(_ vr: VersionCheckResult) -> [String: Any?] {
        switch vr {
        case .forceUpdate(let storePackage):
            return ["result": "forceUpdate", "storePackage": storePackage, "callbackCode": 1]
        case .optionalUpdate(let storePackage):
            return ["result": "optionalUpdate", "storePackage": storePackage, "callbackCode": 2]
        case .noUpdateNeeded:
            return ["result": "noUpdateNeeded", "storePackage": nil, "callbackCode": 0]
        case .error:
            return ["result": "error", "storePackage": nil, "callbackCode": 3]
        }
    }

    private func noticeCheckResultToMap(_ nr: NoticeCheckResult) -> [String: Any?] {
        switch nr {
        case .hasNewNotice(let latestIndex):
            return ["result": "hasNewNotice", "latestIndex": latestIndex]
        case .noNewNotice:
            return ["result": "noNewNotice", "latestIndex": nil]
        case .error:
            return ["result": "error", "latestIndex": nil]
        }
    }
}
